design/vela-cli/sdk_generating.md

```diff
@@ -0,0 +1,114 @@
+2. Storing Application as part of the codebase. Then through some way (e.g. GitOps) the application is applied to cluster.
+
+### Goals & Non-Goals
+
+Goals include the following:
+1. Provide a framework to generate SDK for different languages.
+2. Provide a Go SDK as an example.
+
+Non-goals include the following:
+1. Provide SDK for all languages immediately.
+
+# Proposal
+
+## Benefits
+
+In the aforementioned ways of using Application, there are some issues for both roles.
+
+### For platform team
+
+There are some scenarios which use JSON data passed from frontend is not enough:
+
+1. Platform is responsible for some examination and modification on user's application:
+    1. Adding some traits and policies which is taken care of by platform team. (e.g.  observability)
+    2. Execute some check which can't be done with OpenAPI schema and ui-schema. (e.g. call other services to access needed data)
+2. Platform want to execute some offline batch jobs. For example, add observability trait to existing Applications
+
+In these scenarios, it's [painful](#3606) to develop with `map[string]interface{}`. With SDK, platform team can benefits from the typed X-Definition properties and builds, examines, modifies application more easily.
+
+### For end user
+
+If user is writing Application YAML, they have to refer to the doc. Although we have done some toolchain on this topic like `vela show`, a typed SDK also provide a possibility for end user to code Application in different languages.
+
+## Design and tools
+
+Thanks to the OpenAPI schema and related tool [openapi-generator](https://github.com/OpenAPITools/openapi-generator) in ecosystem, we can re-use the almost ready templates to generate code. The whole generation process is shown below.
+
+![image](../resources/sdk-gen-process.png)
+
+**Step0**: The scaffold is written to destination if `--init` is specified. (omitted in pic)
+
+**Step1**: Generate OpenAPI schema from CUE. The generated schema may not fully prepared as input of openapi-generator and it will be correct and adjust.
+
+**Step2**: Call openapi-generator to generate SDK. After that some modification will be applied to make SDK API suited for KubeVela user.
+
+## Examples
+
+This command requires `java` and `mvn` in PATH as we are leveraging openapi-generator which is a Java progarm.
+
+```shell
+# Initialize the SDK, generate API from all definitions, 
+vela def gen-api -f /path/to/def/dir -o /path/to/sdk --init
+
+# Incrementally generate API from definitions
+vela def gen-api -f /path/to/def/dir -o /path/to/sdk
+```
+
+## Future work
+
+This PR only provide a framework on generating SDK and there are lots of jobs to do to provide
+user-friendly SDKs in multiple languages.
+
+### For SDK generating
+
+1. Provide SDK in other languages. This is a LFX mentorship: #5365
+2. Allow user to customize the templates in Step2.
+
+### For Golang SDK
```

Comment:
We need a section called "Language Caution" to describe the difference and notable points inside.

Reply:
That would be great. And I think it should be part of kubevela.io